File: LaboratoryWork/Semester_3/Laboratory_1/C++/src/ds/DynamicArray.cpp
#include "DynamicArray.hpp"
using namespace std;

void DynamicArray::Init() {
    Size = 0;
    Capacity = 10;
    Data = new string[Capacity];
}

void DynamicArray::PushBack(string obj) {
    if (Size == Capacity) {
        ResizeToRight();
    }
    Data[Size++] = obj;
}

void DynamicArray::Insert(string obj, int pos) {
    if (pos < 0 || pos > Size) {
        throw invalid_argument ("Invalid index!");
    }
    if (Size == Capacity) {
        ResizeToRight();
    }

    for (int i = Size; i > pos; i--) {
        Data[i] = Data[i - 1];
    }
    Data[pos] = obj;
    Size++;
}

string DynamicArray::GetElm(int pos) {
    if (pos < 0 || pos >= Size) {
        throw invalid_argument ("Invalid index!");
    }
    return Data[pos];
}

void DynamicArray::DelElm(int pos) {
    if (pos < 0 || pos >= Size) {
        throw invalid_argument ("Invalid index!");
    }

    for (int i = pos; i < Size - 1; i++) {
        Data[i] = Data[i + 1];
    }

    Size--;
    if (Size == Capacity - 10) {
        ResizeToLeft();
    }
}

void DynamicArray::Replace(string obj, int pos) {
    if (pos < 0 || pos >= Size) {
        throw invalid_argument ("Invalid index!");
    }

    Data[pos] = obj;
}

int DynamicArray::GetSize() {
    return Size;
}

void DynamicArray::Print() {
    for (int i = 0; i < Size; i++) {
        cout << Data[i] << " ";
    }
    cout << endl;
}

void DynamicArray::Destroy() {
    delete[] Data;
    Size = 0;
    Capacity = 0;
}

void DynamicArray::ResizeToRight() {
    int newCapacity = Capacity + 10;
    string* newData = new string[newCapacity];

    for (int i = 0; i < Size; i++) {
        newData[i] = Data[i];
    }

    delete[] Data;
    Data = newData;
    Capacity = newCapacity;
}

void DynamicArray::ResizeToLeft() {
    int newCapacity = Capacity - 10;
    string* newData = new string[newCapacity];

    for (int i = 0; i < Size; i++) {
        newData[i] = Data[i];
    }

    delete[] Data;
    Data = newData;
    Capacity = newCapacity;
}

DynamicArray::~DynamicArray() {
    Destroy();
}

File: LaboratoryWork/Semester_3/Laboratory_2/GoLang/ds/dynamicArray.go
package ds

import (
	"errors"
	"fmt"
)

type DynamicArray[T any] struct {
	data     []T
	size     int
	capacity int
}

func NewDynamicArray[T any]() *DynamicArray[T] {
	capacity := 10
	return &DynamicArray[T]{
		data:     make([]T, capacity),
		size:     0,
		capacity: capacity,
	}
}

func (da *DynamicArray[T]) resizeToRight() {
	newCapacity := da.capacity + 10
	newData := make([]T, newCapacity)
	copy(newData, da.data[:da.size])
	da.data = newData
	da.capacity = newCapacity
}

func (da *DynamicArray[T]) resizeToLeft() {
	if da.capacity <= 10 {
		return
	}
	newCapacity := da.capacity - 10
	newData := make([]T, newCapacity)
	copy(newData, da.data[:da.size])
	da.data = newData
	da.capacity = newCapacity
}

func (da *DynamicArray[T]) PushBack(value T) {
	if da.size == da.capacity {
		da.resizeToRight()
	}
	da.data[da.size] = value
	da.size++
}

func (da *DynamicArray[T]) Insert(value T, pos int) error {
	if pos < 0 || pos > da.size {
		return errors.New("invalid index")
	}
	if da.size == da.capacity {
		da.resizeToRight()
	}
	da.data = append(da.data[:pos+1], da.data[pos:da.size]...)
	da.data[pos] = value
	da.size++
	return nil
}

func (da *DynamicArray[T]) GetElm(pos int) (T, error) {
	if pos < 0 || pos >= da.size {
		var zero T
		return zero, errors.New("invalid index")
	}
	return da.data[pos], nil
}

func (da *DynamicArray[T]) At(pos int) (T, error) {
	return da.GetElm(pos)
}

func (da *DynamicArray[T]) Replace(value T, pos int) error {
	if pos < 0 || pos >= da.size {
		return errors.New("invalid index")
	}
	da.data[pos] = value
	return nil
}

func (da *DynamicArray[T]) DelElm(pos int) error {
	if pos < 0 || pos >= da.size {
		return errors.New("invalid index")
	}
	copy(da.data[pos:], da.data[pos+1:da.size])
	da.size--
	if da.size <= da.capacity-10 {
		da.resizeToLeft()
	}
	return nil
}

func (da *DynamicArray[T]) GetSize() int {
	return da.size
}

func (da *DynamicArray[T]) Clear() {
	da.size = 0
}

func (da *DynamicArray[T]) IsEmpty() bool {
	return da.size == 0
}

func (da *DynamicArray[T]) Destroy() {
	da.data = nil
	da.size = 0
	da.capacity = 0
}

func (da *DynamicArray[T]) Print() {
	fmt.Print("{")
	for i := 0; i < da.size; i++ {
		fmt.Print(da.data[i])
		if i != da.size-1 {
			fmt.Print(", ")
		}
	}
	fmt.Print("}")
}